newIDE/app/src/MainFrame/EditorTabs/DraggableEditorTabs.js

```diff
@@ -0,0 +1,155 @@
+  onClickTab: (index: number) => void,
+  onCloseTab: (editor: EditorTab) => void,
+  onCloseOtherTabs: (editor: EditorTab) => void,
+  onCloseAll: () => void,
+  onTabActive: (editor: EditorTab) => void,
+  onDropTab: (fromIndex: number, toIndex: number) => void,
+|};
+
+export function DraggableClosableTabs({
+  hideLabels,
+  editorTabs,
+  onClickTab,
+  onCloseTab,
+  onCloseOtherTabs,
+  onCloseAll,
+  onTabActive,
+  onDropTab,
+}: DraggableClosableTabsProps) {
+  let draggedTabIndex: number | typeof undefined = undefined;
+
+  return (
+    <ClosableTabs hideLabels={hideLabels}>
+      {getEditors(editorTabs).map((editorTab, id) => {
+        const isCurrentTab = getCurrentTabIndex(editorTabs) === id;
+        return (
+          <DraggableClosableTab
+            index={id}
+            label={editorTab.label}
+            key={editorTab.key}
+            id={`tab-${editorTab.key.replace(/\s/g, '-')}`}
+            active={isCurrentTab}
+            onClick={() => onClickTab(id)}
+            onClose={() => onCloseTab(editorTab)}
+            onCloseOthers={() => onCloseOtherTabs(editorTab)}
+            onCloseAll={onCloseAll}
+            onActivated={() => onTabActive(editorTab)}
+            closable={editorTab.closable}
+            onBeginDrag={() => {
+              draggedTabIndex = id;
+              return editorTab;
+            }}
+            onDrop={toIndex => {
+              if (draggedTabIndex !== undefined) {
+                onDropTab(draggedTabIndex, id);
+                draggedTabIndex = undefined;
+              }
+            }}
+          />
+        );
+      })}
+    </ClosableTabs>
+  );
+}
+
+type DraggableClosableTabProps = {|
+  index: number,
+  onBeginDrag: () => EditorTab,
+  onDrop: (toIndex: number) => void,
+  ...ClosableTabProps,
+|};
+
+export function DraggableClosableTab({
+  index,
+  id,
+  active,
+  onClose,
+  onCloseOthers,
+  onCloseAll,
+  label,
+  closable,
+  onClick,
+  onActivated,
+  onBeginDrag,
+  onDrop,
+}: DraggableClosableTabProps) {
+  const DragSourceAndDropTarget = makeDragSourceAndDropTarget<EditorTab>(
+    'draggable-closable-tab'
+  );
+
+  return (
+    <ScreenTypeMeasurer>
+      {screenType => (
+        <DragSourceAndDropTarget
+          beginDrag={onBeginDrag}
+          canDrag={() => {
+            // On touchscreens, we allow DnD on active tab so that user can scroll
+            if (screenType === 'touch' && !active) return false;
+            // We want "Home" tab to stay on the left
+            return index !== 0;
+          }}
+          canDrop={() => true}
+          drop={() => onDrop(index)}
+        >
+          {({ connectDragSource, connectDropTarget, isOver, canDrop }) => {
+            // If on a touch screen, setting the whole item to be
+            // draggable would prevent scroll. Set the icon only to be
+            // draggable if the item is not selected. When selected,
+            // set the whole item to be draggable.
```

Comment:
Maybe the hook `useLongTouch` can be of use here. I've not used it so far so I'm not sure, but if you want to improve the mobile user experience, it could be handy.

Reply:
I agree, using long touch is a good idea. (And now that I tried it, allowing active tab to move on drag makes it become very slippery since you could try to scroll without paying too much attention to where you start scrolling...)

But long touch is already used to display the context menu, so I am not sure how to trigger one but not the other ? :/